pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,835 @@
+	// a vpa-admission-controller deployment, and disable it otherwise
+	if spec.Settings.VerticalPodAutoscaler == nil {
+		disabled, err := a.disableVPA(ctx, shoot)
+		if err != nil {
+			return nil, err
+		}
+		spec.Settings.VerticalPodAutoscaler = &gardencorev1beta1.SeedSettingVerticalPodAutoscaler{
+			Enabled: !disabled,
+		}
+	}
+
+	return spec, nil
+}
+
+func (a *actuator) prepareSeedBackup(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot, shootSecret *corev1.Secret, seedBackup *gardencorev1beta1.SeedBackup) (*gardencorev1beta1.SeedBackup, error) {
+	if seedBackup == nil {
+		return nil, nil
+	}
+
+	// Copy backup
+	backup := seedBackup.DeepCopy()
+
+	// Initialize provider
+	if backup.Provider == "" {
+		backup.Provider = shoot.Spec.Provider.Type
+	}
+
+	// If backup secret reference is not specified, initialize it and create or update the corresponding secret
+	if backup.SecretRef.Name == "" || backup.SecretRef.Namespace == "" {
+		// Initialize backup secret reference
+		backup.SecretRef = corev1.SecretReference{
+			Name:      fmt.Sprintf("backup-%s", managedSeed.Name),
+			Namespace: managedSeed.Namespace,
+		}
+
+		// Create or update backup secret
+		ownerRefs := []metav1.OwnerReference{
+			*metav1.NewControllerRef(managedSeed, seedmanagementv1alpha1.SchemeGroupVersion.WithKind("ManagedSeed")),
+		}
+		if err := kutil.CreateOrUpdateSecretByRef(ctx, a.gardenClient.Client(), backup.SecretRef, corev1.SecretTypeOpaque, shootSecret.Data, ownerRefs); err != nil {
+			return nil, err
+		}
+	}
+
+	return backup, nil
+}
+
+func (a *actuator) getShootSecret(ctx context.Context, shoot *gardencorev1beta1.Shoot) (*corev1.Secret, error) {
+	shootSecretBinding := &gardencorev1beta1.SecretBinding{}
+	if err := a.gardenClient.Client().Get(ctx, kutil.Key(shoot.Namespace, shoot.Spec.SecretBindingName), shootSecretBinding); err != nil {
+		return nil, err
+	}
+	return kutil.GetSecretByRef(ctx, a.gardenClient.Client(), shootSecretBinding.SecretRef)
+}
+
+func (a *actuator) getShootKubeconfigSecret(ctx context.Context, shoot *gardencorev1beta1.Shoot) (*corev1.Secret, error) {
+	shootKubeconfigSecret := &corev1.Secret{}
+	if err := a.gardenClient.Client().Get(ctx, kutil.Key(shoot.Namespace, fmt.Sprintf("%s.kubeconfig", shoot.Name)), shootKubeconfigSecret); err != nil {
+		return nil, err
+	}
+	return shootKubeconfigSecret, nil
+}
+
+func (a *actuator) disableVPA(ctx context.Context, shoot *gardencorev1beta1.Shoot) (bool, error) {
+	if err := a.seedClient.Client().Get(ctx, kutil.Key(shoot.Status.TechnicalID, "vpa-admission-controller"), &appsv1.Deployment{}); err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+const (
+	gardenletKubeconfigBootstrapSecretName = "gardenlet-kubeconfig-bootstrap"
+	gardenletKubeconfigSecretName          = "gardenlet-kubeconfig"
```

Comment:
Hm, this is a good question. I actually had similar thoughts but decided against it after observing how it actually works. `bootstrapKubeconfig` is set only if the bootstrap didn't happen yet, in order to make it happen, otherwise it's cleared (it would actually confuse gardenlet to keep it). Therefore, even if it's specified, it will be cleared almost every single time. I think that the user shouldn't care about the name of this secret and actually should not try to specify `bootstrapKubeconfig` at all. For consistency, the same follows for `kubeconfigSecret`, it's another secret that is created by the implementation and therefore the user shouldn't really care about its name.

Another reason to hardcode these secrets is that they have to be cleaned up at some point. With hardcoded names it's easy, otherwise we would need to check if the user specified any names, and use the default names if he didn't specify, etc, which makes it much more complicated for no good reason.

Of course, it needs to be documented and validated that the user didn't actually specify anything there.

Reply:
> bootstrapKubeconfig is set only if the bootstrap didn't happen yet, in order to make it happen, otherwise it's cleared (it would actually confuse gardenlet to keep it). 

Do you mean the Gardenelt configuration `bootstrapKubeconfig` or the secret that it references? Because the configuration is not deleted by the Gardenlet, only the secret behind it. This is done for security purposes. It would not confuse the Gardenlet if the bootstrap secret is there. If the secret configured in kubeconfigSecret exists, [then it simply uses that and skips the bootstrap](https://github.com/gardener/gardener/blob/03bf8dcebb2032f297a4f40b90236c6e759bcca9/cmd/gardenlet/app/bootstrap.go#L47).

> Therefore, even if it's specified, it will be cleared almost every single time. 

In my point of view, this is not an argument to forbid to specify alternative secret names. 
I think it can be handy if you want to provide a pre-provisioned bootstrap kubeconfig.
Atm, I do not see a real reason why we would restrict the secret names to only those constants. Especially, because the Gardenlet configuration allows it.

>  it's another secret that is created by the implementation and therefore the user shouldn't really care about its name.

While they do not really need to care, I think they should still be allowed to specify another secret name?

> Another reason to hardcode these secrets is that they have to be cleaned up at some point. With hardcoded names it's easy, otherwise, we would need to check if the user has specified any names, and use the default names if he didn't specify, etc, which makes it much more complicated for no good reason.

Does not seem like much of an effort to me.

Reply:
Thinking about it again, I do not have a strong opinion here. I am also fine if we just validation and documentation. 
Maybe others have an opinion here as well.

Reply:
I was wrong about cleanup, I had in mind an old implementation that I subsequently refactored, so this is not a consideration. I will check how hard it would be to use the names that the user specified and the constants above only as defaults if nothing was specified, and I see no major challenge I will do it this way.

Reply:
OK, I found a way to do it without adding too much complexity.